File: packages/provide-uterm-go/ctrlmsg/patterns.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package ctrlmsg

import (
	"fmt"
	"strings"
	"sync"
)

// LinkPattern is an immutable descriptor for one server-driven clickable text
// pattern, ported from the Python control_channel_patterns.LinkPattern
// dataclass. Construct one with NewLinkPattern so the action is validated and
// the "g" flags default is applied.
type LinkPattern struct {
	// Pattern is the JavaScript regex source string.
	Pattern string
	// Action is what happens on click: "cmd", "url", "key" or "focus".
	Action string
	// ID is the optional stable identifier used by the registry (nil == none).
	ID *string
	// Flags are the regex flags forwarded to new RegExp; defaults to "g".
	Flags string
	// Group is the clickable capture group (0 == whole match).
	Group int
	// Payload is the click payload template.
	Payload string
	// Hover is the hover-tooltip template.
	Hover string
	// Class is the CSS class applied to highlighted ranges; serialised as the
	// wire key "class".
	Class string
}

// LinkPatternOption configures NewLinkPattern.
type LinkPatternOption func(*LinkPattern)

// WithID sets the stable identifier used for registry replace/unregister.
func WithID(id string) LinkPatternOption {
	return func(p *LinkPattern) { p.ID = &id }
}

// WithFlags overrides the regex flags (default "g").
func WithFlags(flags string) LinkPatternOption {
	return func(p *LinkPattern) { p.Flags = flags }
}

// WithGroup sets the clickable capture group.
func WithGroup(group int) LinkPatternOption {
	return func(p *LinkPattern) { p.Group = group }
}

// WithPayload sets the click payload template.
func WithPayload(payload string) LinkPatternOption {
	return func(p *LinkPattern) { p.Payload = payload }
}

// WithHover sets the hover-tooltip template.
func WithHover(hover string) LinkPatternOption {
	return func(p *LinkPattern) { p.Hover = hover }
}

// WithClass sets the CSS class applied to highlighted ranges.
func WithClass(class string) LinkPatternOption {
	return func(p *LinkPattern) { p.Class = class }
}

// NewLinkPattern builds a LinkPattern for the given regex source and action,
// applying the "g" flags default and validating the action. It mirrors the
// Python dataclass __post_init__ validation; an unknown action is an error.
//
// The Python error lists the valid actions as the sorted repr
// ['cmd', 'focus', 'key', 'url']; this Go port emits the same content with Go
// slice formatting: `invalid action "explode"; must be one of [cmd focus key url]`.
func NewLinkPattern(pattern, action string, opts ...LinkPatternOption) (LinkPattern, error) {
	p := LinkPattern{Pattern: pattern, Action: action, Flags: "g"}
	for _, opt := range opts {
		opt(&p)
	}
	if !isValidLinkAction(p.Action) {
		return LinkPattern{}, fmt.Errorf(
			"invalid action %q; must be one of [%s]", p.Action, strings.Join(validLinkActions, " "),
		)
	}
	return p, nil
}

// ToFrameEntry serialises the pattern to its wire-format map. Only non-default
// optional fields are included: "flags" is omitted when "g", "group" when 0,
// and "id"/"payload"/"hover"/"class" when empty. Class is emitted under the key
// "class".
func (p LinkPattern) ToFrameEntry() map[string]any {
	entry := map[string]any{
		"pattern": p.Pattern,
		"action":  p.Action,
	}
	if p.ID != nil {
		entry["id"] = *p.ID
	}
	if p.Flags != "g" {
		entry["flags"] = p.Flags
	}
	if p.Group != 0 {
		entry["group"] = p.Group
	}
	if p.Payload != "" {
		entry["payload"] = p.Payload
	}
	if p.Hover != "" {
		entry["hover"] = p.Hover
	}
	if p.Class != "" {
		entry["class"] = p.Class
	}
	return entry
}

// LinkPatternRegistry tracks the active link-pattern set for one owner
// (session, worker, etc.), ported from the Python LinkPatternRegistry.
//
// Patterns are kept in insertion order. Registering a pattern whose ID already
// exists replaces the earlier one in place (its slot position is preserved).
// Patterns without an ID are appended and cannot be removed individually.
//
// Unlike the Python original — which relies on the GIL for atomicity — this
// port guards its state with a sync.Mutex, so it is safe for concurrent use.
type LinkPatternRegistry struct {
	mu      sync.Mutex
	order   []any // key insertion order: string ID or int counter sentinel
	items   map[any]LinkPattern
	counter int
}

// NewLinkPatternRegistry returns an empty registry ready for use. The zero
// value is also usable; this constructor is provided for clarity.
func NewLinkPatternRegistry() *LinkPatternRegistry {
	return &LinkPatternRegistry{}
}

func (r *LinkPatternRegistry) ensureInit() {
	if r.items == nil {
		r.items = make(map[any]LinkPattern)
	}
}

// Register adds pattern to the active set. A pattern with an ID that already
// exists replaces the earlier entry in place; an ID-less pattern is appended
// under a private counter key.
func (r *LinkPatternRegistry) Register(pattern LinkPattern) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.ensureInit()

	var key any
	if pattern.ID != nil {
		key = *pattern.ID
	} else {
		key = r.counter
		r.counter++
	}
	if _, exists := r.items[key]; !exists {
		r.order = append(r.order, key)
	}
	r.items[key] = pattern
}

// Unregister removes the pattern registered under patternID, returning true if
// one was found and removed. ID-less patterns (keyed by an int counter) never
// match a string ID.
func (r *LinkPatternRegistry) Unregister(patternID string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.ensureInit()

	if _, exists := r.items[patternID]; !exists {
		return false
	}
	delete(r.items, patternID)
	for i, k := range r.order {
		if k == patternID {
			r.order = append(r.order[:i], r.order[i+1:]...)
			break
		}
	}
	return true
}

// Clear removes all patterns and resets the ID-less counter.
func (r *LinkPatternRegistry) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.items = make(map[any]LinkPattern)
	r.order = nil
	r.counter = 0
}

// GetAll returns all active patterns in insertion order.
func (r *LinkPatternRegistry) GetAll() []LinkPattern {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]LinkPattern, 0, len(r.order))
	for _, k := range r.order {
		out = append(out, r.items[k])
	}
	return out
}

// SyncPayload returns a fresh {"type": "link_patterns", "patterns": [...]} map
// ready for controlchannel.EncodeControlFrame. It is non-destructive.
func (r *LinkPatternRegistry) SyncPayload() map[string]any {
	r.mu.Lock()
	defer r.mu.Unlock()
	patterns := make([]any, 0, len(r.order))
	for _, k := range r.order {
		patterns = append(patterns, r.items[k].ToFrameEntry())
	}
	return map[string]any{"type": "link_patterns", "patterns": patterns}
}

File: packages/provide-uterm-ts/src/ctrlmsg/patterns.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Registry for `link_patterns` control-channel frames.
 *
 * {@link LinkPattern} is an immutable value object describing one
 * server-driven clickable text decoration. {@link LinkPatternRegistry} tracks
 * the active set for a single owner (session, worker, and so on); callers
 * create one registry per owner and there is no shared global state.
 *
 * Port of the Python module `provide.uterm.control_channel_patterns`.
 */

/** The four things a click can do. */
export type LinkPatternAction = "cmd" | "url" | "key" | "focus";

/** Valid actions, for validation and for the error message. */
const VALID_ACTIONS: readonly LinkPatternAction[] = ["cmd", "focus", "key", "url"];

/** Construction fields for {@link LinkPattern}. */
export interface LinkPatternInit {
  /** JavaScript regex source string. */
  pattern: string;
  /** What happens when the user clicks. */
  action: LinkPatternAction;
  /**
   * Stable identifier used by {@link LinkPatternRegistry.unregister}. When two
   * patterns share an id the later `register` replaces the earlier one.
   */
  id?: string;
  /**
   * Regex flags forwarded to `new RegExp(pattern, flags)`. Defaults to `"g"`;
   * the client ensures `"g"` regardless.
   */
  flags?: string;
  /** Which capture group is the clickable span; 0 is the whole match. */
  group?: number;
  /** Click payload template, with `$1`, `$2`, … substituted from captures. */
  payload?: string;
  /** Hover tooltip template, with the same substitution. */
  hover?: string;
  /**
   * CSS class applied to highlighted ranges. Serialised as `"class"`, which
   * the Python side spells `class_` because `class` is reserved there.
   */
  className?: string;
}

/** An immutable descriptor for one server-driven clickable text pattern. */
export class LinkPattern {
  readonly pattern: string;
  readonly action: LinkPatternAction;
  readonly id: string | undefined;
  readonly flags: string;
  readonly group: number;
  readonly payload: string;
  readonly hover: string;
  readonly className: string;

  constructor(init: LinkPatternInit) {
    if (!VALID_ACTIONS.includes(init.action)) {
      throw new Error(`invalid action ${JSON.stringify(init.action)}; must be one of ${JSON.stringify(VALID_ACTIONS)}`);
    }
    this.pattern = init.pattern;
    this.action = init.action;
    this.id = init.id;
    this.flags = init.flags ?? "g";
    this.group = init.group ?? 0;
    this.payload = init.payload ?? "";
    this.hover = init.hover ?? "";
    this.className = init.className ?? "";
  }

  /**
   * Serialise to the wire-format entry the browser expects.
   *
   * Only non-default and non-empty optional fields are included, to keep
   * frames compact.
   */
  toFrameEntry(): Record<string, unknown> {
    const entry: Record<string, unknown> = { pattern: this.pattern, action: this.action };
    if (this.id !== undefined) {
      entry.id = this.id;
    }
    if (this.flags !== "g") {
      entry.flags = this.flags;
    }
    if (this.group !== 0) {
      entry.group = this.group;
    }
    if (this.payload !== "") {
      entry.payload = this.payload;
    }
    if (this.hover !== "") {
      entry.hover = this.hover;
    }
    if (this.className !== "") {
      entry.class = this.className;
    }
    return entry;
  }
}

/**
 * Active pattern set for one owner.
 *
 * Patterns are stored in insertion order. Registering a pattern whose id is
 * already present replaces the earlier one in its existing slot, so order
 * stays predictable for callers that register once and later refresh the same
 * id. Patterns without an id are appended and cannot be removed individually;
 * use {@link clear} to reset the whole set.
 *
 * A `Map` keyed by id or by a monotonic counter reproduces the reference's
 * `dict` semantics exactly: `Map` preserves insertion order for every key
 * type, where a plain object would reorder the integer-like counter keys
 * ahead of the string ids.
 */
export class LinkPatternRegistry {
  readonly #patterns = new Map<string | number, LinkPattern>();
  #counter = 0;

  /** Add `pattern` to the active set. */
  register(pattern: LinkPattern): void {
    if (pattern.id !== undefined) {
      this.#patterns.set(pattern.id, pattern);
      return;
    }
    this.#patterns.set(this.#counter, pattern);
    this.#counter += 1;
  }

  /**
   * Remove the pattern registered under `id`.
   *
   * @returns `true` when a pattern was found and removed.
   */
  unregister(id: string): boolean {
    return this.#patterns.delete(id);
  }

  /** Remove all patterns and reset the id-less counter. */
  clear(): void {
    this.#patterns.clear();
    this.#counter = 0;
  }

  /** All active patterns, in insertion order. */
  getAll(): LinkPattern[] {
    return [...this.#patterns.values()];
  }

  /**
   * The ready-to-send payload for `encodeControlFrame`.
   *
   * Non-destructive: the registry is unchanged.
   */
  syncPayload(): Record<string, unknown> {
    return {
      type: "link_patterns",
      patterns: this.getAll().map((pattern) => pattern.toFrameEntry()),
    };
  }
}
